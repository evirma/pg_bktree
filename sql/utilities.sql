-- int8 check


-- SELECT pg_sleep(20);

SELECT int64_to_bitstring(0);
SELECT int64_to_bitstring(1);
SELECT int64_to_bitstring(2);
SELECT int64_to_bitstring(-1);
SELECT int64_to_bitstring(-2);
SELECT int64_to_bitstring(9223372036854775807);

SELECT bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000');
SELECT bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000001');
SELECT bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111');
SELECT bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111');
SELECT bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111110');

-- Check function round-tripping

-- Annoying casting is because I haven't figured out how to return an instance of "text" from a c function
-- Mostly because it works as is, and it's not critical.
SELECT int64_to_bitstring(bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000'))::text
	= '0000000000000000000000000000000000000000000000000000000000000000'::text;
SELECT int64_to_bitstring(bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000001'))::text
	= '0000000000000000000000000000000000000000000000000000000000000001'::text;
SELECT int64_to_bitstring(bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111'))::text
	= '0111111111111111111111111111111111111111111111111111111111111111'::text;
SELECT int64_to_bitstring(bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111'))::text
	= '1111111111111111111111111111111111111111111111111111111111111111'::text;
SELECT int64_to_bitstring(bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111110'))::text
	= '1111111111111111111111111111111111111111111111111111111111111110'::text;


SELECT bitstring_to_int64(int64_to_bitstring(0)) = 0;
SELECT bitstring_to_int64(int64_to_bitstring(1)) = 1;
SELECT bitstring_to_int64(int64_to_bitstring(2)) = 2;
SELECT bitstring_to_int64(int64_to_bitstring(-1)) = -1;
SELECT bitstring_to_int64(int64_to_bitstring(-2)) = -2;
SELECT bitstring_to_int64(int64_to_bitstring(9223372036854775807)) = 9223372036854775807;

-- We expect these to fail
SELECT bitstring_to_int64('01111111111111111111111111111111111111111111111111111111111111110');

SELECT bitstring_to_int64('111111111111111111111111111111111111111111111111111111111111110');

SELECT bitstring_to_int64('11111111111111111111xx111111111111111111111111111111111111111110');

SELECT bitstring_to_int64('wat');


-- Validate the hamming distance operator
-- A lot of this is trying to break anything that depends on the
-- sign bit when it shouldn't. Some of it is just silly, some of
-- it tries to validate the minimum and maximum distances.

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
;

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000001')
;

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000000')
;

SELECT
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000000')
;

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000001')
;

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000011')
;

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111')
;

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;

SELECT
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111110')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;

SELECT
		bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;

SELECT
		bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111110')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;
