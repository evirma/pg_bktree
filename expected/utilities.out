-- int8 check
-- SELECT pg_sleep(20);
SELECT int64_to_bitstring(0);
                        int64_to_bitstring                        
------------------------------------------------------------------
 0000000000000000000000000000000000000000000000000000000000000000
(1 row)

SELECT int64_to_bitstring(1);
                        int64_to_bitstring                        
------------------------------------------------------------------
 0000000000000000000000000000000000000000000000000000000000000001
(1 row)

SELECT int64_to_bitstring(2);
                        int64_to_bitstring                        
------------------------------------------------------------------
 0000000000000000000000000000000000000000000000000000000000000010
(1 row)

SELECT int64_to_bitstring(-1);
                        int64_to_bitstring                        
------------------------------------------------------------------
 1111111111111111111111111111111111111111111111111111111111111111
(1 row)

SELECT int64_to_bitstring(-2);
                        int64_to_bitstring                        
------------------------------------------------------------------
 1111111111111111111111111111111111111111111111111111111111111110
(1 row)

SELECT int64_to_bitstring(9223372036854775807);
                        int64_to_bitstring                        
------------------------------------------------------------------
 0111111111111111111111111111111111111111111111111111111111111111
(1 row)

SELECT bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000');
 bitstring_to_int64 
--------------------
                  0
(1 row)

SELECT bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000001');
 bitstring_to_int64 
--------------------
                  1
(1 row)

SELECT bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111');
 bitstring_to_int64  
---------------------
 9223372036854775807
(1 row)

SELECT bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111');
 bitstring_to_int64 
--------------------
                 -1
(1 row)

SELECT bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111110');
 bitstring_to_int64 
--------------------
                 -2
(1 row)

-- Check function round-tripping
-- Annoying casting is because I haven't figured out how to return an instance of "text" from a c function
-- Mostly because it works as is, and it's not critical.
SELECT int64_to_bitstring(bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000'))::text
	= '0000000000000000000000000000000000000000000000000000000000000000'::text;
 ?column? 
----------
 t
(1 row)

SELECT int64_to_bitstring(bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000001'))::text
	= '0000000000000000000000000000000000000000000000000000000000000001'::text;
 ?column? 
----------
 t
(1 row)

SELECT int64_to_bitstring(bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111'))::text
	= '0111111111111111111111111111111111111111111111111111111111111111'::text;
 ?column? 
----------
 t
(1 row)

SELECT int64_to_bitstring(bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111'))::text
	= '1111111111111111111111111111111111111111111111111111111111111111'::text;
 ?column? 
----------
 t
(1 row)

SELECT int64_to_bitstring(bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111110'))::text
	= '1111111111111111111111111111111111111111111111111111111111111110'::text;
 ?column? 
----------
 t
(1 row)

SELECT bitstring_to_int64(int64_to_bitstring(0)) = 0;
 ?column? 
----------
 t
(1 row)

SELECT bitstring_to_int64(int64_to_bitstring(1)) = 1;
 ?column? 
----------
 t
(1 row)

SELECT bitstring_to_int64(int64_to_bitstring(2)) = 2;
 ?column? 
----------
 t
(1 row)

SELECT bitstring_to_int64(int64_to_bitstring(-1)) = -1;
 ?column? 
----------
 t
(1 row)

SELECT bitstring_to_int64(int64_to_bitstring(-2)) = -2;
 ?column? 
----------
 t
(1 row)

SELECT bitstring_to_int64(int64_to_bitstring(9223372036854775807)) = 9223372036854775807;
 ?column? 
----------
 t
(1 row)

-- We expect these to fail
SELECT bitstring_to_int64('01111111111111111111111111111111111111111111111111111111111111110');
ERROR:  bitstring_to_int64() input MUST be 64 characters long. Passed string length: 65

SELECT bitstring_to_int64('111111111111111111111111111111111111111111111111111111111111110');
ERROR:  bitstring_to_int64() input MUST be 64 characters long. Passed string length: 63

SELECT bitstring_to_int64('11111111111111111111xx111111111111111111111111111111111111111110');
ERROR:  bitstring_to_int64() input can only contain either '1' or '0'. Found bad char 'x' at index 20

SELECT bitstring_to_int64('wat');
ERROR:  bitstring_to_int64() input MUST be 64 characters long. Passed string length: 3

-- Validate the hamming distance operator
-- A lot of this is trying to break anything that depends on the
-- sign bit when it shouldn't. Some of it is just silly, some of
-- it tries to validate the minimum and maximum distances.
SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
;
 ?column? 
----------
        0
(1 row)

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000001')
;
 ?column? 
----------
        1
(1 row)

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000000')
;
 ?column? 
----------
        1
(1 row)

SELECT
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000000')
;
 ?column? 
----------
        0
(1 row)

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1000000000000000000000000000000000000000000000000000000000000001')
;
 ?column? 
----------
        2
(1 row)

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000011')
;
 ?column? 
----------
        2
(1 row)

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111')
;
 ?column? 
----------
       63
(1 row)

SELECT
		bitstring_to_int64('0000000000000000000000000000000000000000000000000000000000000000')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;
 ?column? 
----------
       64
(1 row)

SELECT
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111110')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;
 ?column? 
----------
        1
(1 row)

SELECT
		bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111111')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;
 ?column? 
----------
        1
(1 row)

SELECT
		bitstring_to_int64('0111111111111111111111111111111111111111111111111111111111111110')
	<->
		bitstring_to_int64('1111111111111111111111111111111111111111111111111111111111111111')
;
 ?column? 
----------
        2
(1 row)

